### PR TITLE
Annotation methods

### PR DESCRIPTION
**Pull request by philipjackson**
# Description

Since we need to log every change to the `annotationsObject`, we need all changes to be done via methods on that object which can log their effects. Previously, particularly in `SplineCanvas`, we were making most of our changes by getting data structures out of `annotationsObject` by reference (e.g. `const { coordinates } = this.props.annotationsObject.getActiveAnnotation();`), and modifying them directly. If we kept doing that then we'd have to manually log all of these changes and it'd get messy and there would be bugs. So I've added a bunch of methods to the `Annotations` class like `addSplinePoint` and `addBrushStroke`, and removed methods like `getActiveAnnotation` where possible (there are a couple of methods like that that I couldn't remove, so instead I had them return a copy of a data structure rather than a reference to it, so they can't be used to modify the internal state of `annotationsObject`).

I think I've fixed all the bugs this introduced, but if someone could mess around with the paintbrush and spline tools and try to break them, that'd be much appreciated.

relates #151 

# Dependency changes

N/A

# Testing

N/A

# Documentation

N/A

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] New migrations have been committed
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by ChasNelson1990** [comment]
Broke it! Tried to magic spline. Clicked 'magic spline', clicked 'new annotation', tried to spline

Console Logs:

```
11:21:54.940 contentscript.js:24 contentscript.js: https://black-wave-0f4ddd903-159.westeurope.azurestaticapps.net/ 
11:21:56.073 /favicon.ico:1 Failed to load resource: the server responded with a status of 404 ()
11:25:26.649 Canvas.tsx:141 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.650 react-dom.development.js:4091 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.652 Canvas.tsx:141 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.652 react-dom.development.js:4091 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.670 Canvas.tsx:141 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.670 react-dom.development.js:4091 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.687 Canvas.tsx:141 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.687 react-dom.development.js:4091 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.705 Canvas.tsx:484 Uncaught TypeError: Cannot read property 'x' of null
    at SplineCanvas.snapToGradient (Canvas.tsx:484)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:571)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
    at executeDispatch (react-dom.development.js:8243)
    at processDispatchQueueItemsInOrder (react-dom.development.js:8275)
    at processDispatchQueue (react-dom.development.js:8288)
11:25:26.705 react-dom.development.js:4091 Uncaught TypeError: Cannot read property 'x' of null
    at SplineCanvas.snapToGradient (Canvas.tsx:484)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:571)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
    at executeDispatch (react-dom.development.js:8243)
    at processDispatchQueueItemsInOrder (react-dom.development.js:8275)
    at processDispatchQueue (react-dom.development.js:8288)
11:25:26.720 Canvas.tsx:141 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.720 react-dom.development.js:4091 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.737 Canvas.tsx:141 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.737 react-dom.development.js:4091 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.753 Canvas.tsx:141 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.754 react-dom.development.js:4091 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.770 Canvas.tsx:141 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.770 react-dom.development.js:4091 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.789 Canvas.tsx:141 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.789 react-dom.development.js:4091 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.831 Canvas.tsx:141 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.831 react-dom.development.js:4091 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.839 Canvas.tsx:141 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.839 react-dom.development.js:4091 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.854 Canvas.tsx:141 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.854 react-dom.development.js:4091 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.871 Canvas.tsx:141 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.871 react-dom.development.js:4091 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.890 Canvas.tsx:141 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.890 react-dom.development.js:4091 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at Object.SplineCanvas.onMouseMove (Canvas.tsx:593)
    at BaseCanvas.onMouseMoveHandler (Canvas.tsx:128)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
11:25:26.903 Canvas.tsx:283 Uncaught TypeError: Cannot read property 'x' of null
    at SplineCanvas.clickNearPoint (Canvas.tsx:283)
    at Object.SplineCanvas.onClick (Canvas.tsx:316)
    at BaseCanvas.onClickHandler (Canvas.tsx:112)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
    at executeDispatch (react-dom.development.js:8243)
    at processDispatchQueueItemsInOrder (react-dom.development.js:8275)
    at processDispatchQueue (react-dom.development.js:8288)
11:25:26.903 react-dom.development.js:4091 Uncaught TypeError: Cannot read property 'x' of null
    at SplineCanvas.clickNearPoint (Canvas.tsx:283)
    at Object.SplineCanvas.onClick (Canvas.tsx:316)
    at BaseCanvas.onClickHandler (Canvas.tsx:112)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070)
    at executeDispatch (react-dom.development.js:8243)
    at processDispatchQueueItemsInOrder (react-dom.development.js:8275)
    at processDispatchQueue (react-dom.development.js:8288)
11:25:27.684 Canvas.tsx:141 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at SplineCanvas.componentDidUpdate (Canvas.tsx:86)
    at commitLifeCycles (react-dom.development.js:20684)
    at commitLayoutEffects (react-dom.development.js:23426)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
11:25:27.696 react-dom.development.js:20085 The above error occurred in the <SplineCanvas> component:

    at SplineCanvas (https://black-wave-0f4ddd903-159.westeurope.azurestaticapps.net/main.js:65522:9)
    at div
    at Grid (https://black-wave-0f4ddd903-159.westeurope.azurestaticapps.net/main.js:5099:35)
    at WithStyles (https://black-wave-0f4ddd903-159.westeurope.azurestaticapps.net/main.js:17082:31)
    at div
    at Grid (https://black-wave-0f4ddd903-159.westeurope.azurestaticapps.net/main.js:5099:35)
    at WithStyles (https://black-wave-0f4ddd903-159.westeurope.azurestaticapps.net/main.js:17082:31)
    at div
    at Container (https://black-wave-0f4ddd903-159.westeurope.azurestaticapps.net/main.js:3824:23)
    at WithStyles (https://black-wave-0f4ddd903-159.westeurope.azurestaticapps.net/main.js:17082:31)
    at ThemeProvider (https://black-wave-0f4ddd903-159.westeurope.azurestaticapps.net/main.js:16244:24)
    at UserInterface (https://black-wave-0f4ddd903-159.westeurope.azurestaticapps.net/main.js:66298:9)

Consider adding an error boundary to your tree to customize error handling behavior.
Visit https://reactjs.org/link/error-boundaries to learn more about error boundaries.
logCapturedError @ react-dom.development.js:20085
11:25:27.696 react-dom.development.js:11340 Uncaught TypeError: Cannot destructure property 'x' of '.for' as it is undefined.
    at SplineCanvas.drawSplineVector (Canvas.tsx:141)
    at Canvas.tsx:200
    at Array.forEach (<anonymous>)
    at SplineCanvas.drawAllSplines (Canvas.tsx:199)
    at SplineCanvas.componentDidUpdate (Canvas.tsx:86)
    at commitLifeCycles (react-dom.development.js:20684)
    at commitLayoutEffects (react-dom.development.js:23426)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
11:25:30.515 DevTools failed to load SourceMap: Could not load content for chrome-extension://ekiibapogjgmlhlhpoalbppfhhgkcogc/qtip/jquery.qtip.min.map: HTTP error: status code 404, net::ERR_UNKNOWN_URL_SCHEME

```

File: src/annotation/index.ts [hunk 1]
```diff
@@ -30,6 +30,63 @@ export class Annotations {
...
+    this.data.splice(this.activeAnnotationID, 1);
+    if (this.activeAnnotationID >= this.data.length) {
+      this.activeAnnotationID = this.data.length - 1; // necessary if we delete the one on the end
+    }
+    if (this.data.length === 0) {
+      this.addAnnotation("paintbrush"); // re-create a new empty annotation if we delete the last one (toolbox will be re-assigned by reuseEmptyAnnotation if necessary)
+    }
+  };
+
+  getLabels = (): string[] => this.data[this.activeAnnotationID].labels;
```
Comment (ChasNelson1990) [inline src/annotation/index.ts:43]: Should we use CRUD terminology here? I.e. createLabels, readLabels, updateLabels, deleteLabels?

Comment (ChasNelson1990) [reply src/annotation/index.ts:43]: We currently have: delete, get, add, clear, update, insert, and set

**Review by ChasNelson1990 — APPROVED** [review]

**Review by ChasNelson1990 — DISMISSED** [review]

File: src/annotation/index.ts (hunk 1 above)
Comment (cooper667) [reply src/annotation/index.ts:43]: I tend to use get/create fwiw

Comment (cooper667) [reply src/annotation/index.ts:43]: image

**Review by SilviaZeta — APPROVED** [review]

**Review by ChasNelson1990 — APPROVED** [review]

**Pull request merged**